### ISSUE TITLE
Cross database foreign key support

### ISSUE THREAD

**Issue by stooie90** [miguelgrinberg/Flask-Migrate#91]
I know Flask-Migrate supports multiple databases using the Flask-SQLAlchemy binds however, I'm unsure if it supports ForeignKeys across those databases (specifically using MySQL). Take an extremely basic example:

```
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_script import Manager
from flask_migrate import Migrate, MigrateCommand

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = "mysql://root:@localhost/db1"
app.config['SQLALCHEMY_BINDS'] = {
    'db2': 'mysql://root@localhost/db2'
}

db = SQLAlchemy(app)
migrate = Migrate(app, db)
manager = Manager(app)
manager.add_command('db', MigrateCommand)

class T1(db.Model):
    __tablename__ = "t1"
    t1_id = db.Column(db.Integer(), primary_key=True)
    t2_id = db.Column(db.Integer(), db.ForeignKey('t2.t2_id'))

class T2(db.Model):
    __bind_key__ = 'db2'
    __tablename__ = "t1"
    t2_id = db.Column(db.Integer(), primary_key=True)

if __name__ == '__main__':
    manager.run()
```

When I run the following:

```
python manage.py db init --multidb
python manage.py db migrate
```

I end up with an error saying that I cannot find the table I'm referencing which is located in another database.

`sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 't1.t2_id' could not find table 't2' with which to generate a foreign key to target column 't2_id'`

I'm pretty sure that I can create a blank revision and do this manually but it'd be awesome if the `migrate` command could pick up on this.

**Comment by miguelgrinberg** [comment]
This would be a feature request for Alembic. You can write it on that project: https://bitbucket.org/zzzeek/alembic.

**Comment by stooie90** [comment]
Turned out it wasn't really an Alembic issue, it was more of an issue of how Flask-SQLAlchemy handles its binds. There is some magic there with allows Flask-SQLAlchemy to work out the schema the table belongs to but SQLAlchemy + Alembic require the schema name to be explicit. I came up with this work around.

    from flask import Flask, current_app
    from flask_sqlalchemy import SQLAlchemy
    from flask_script import Manager
    from flask_migrate import Migrate, MigrateCommand
    from sqlalchemy.ext.declarative import declared_attr
    from urlparse import urlparse
    
    app = Flask(__name__)
    app.config['SQLALCHEMY_DATABASE_URI'] = "mysql://root:@localhost/db1"
    app.config['SQLALCHEMY_BINDS'] = {
        'db2': 'mysql://root@localhost/db2'
    }
    
    # When trying to find the schema names dynamically, the models need access to the app.config
    # so I need to push the context manually since current_app is only initialised in the flask
    # request context. [trivial for this example but required if no reference to app]
    app_context = app.app_context()
    app_context.push()
    
    db = SQLAlchemy(app)
    migrate = Migrate(app, db)
    manager = Manager(app)
    manager.add_command('db', MigrateCommand)
    
    
    def create_fk(column, schema=None, **kwargs):
        if schema:
            column = "%s.%s" % (schema, column)
        return db.ForeignKey(column, **kwargs)
    
    
    def _find_schema(cls):
        if hasattr(cls, '__bind_key__'):
            binds = current_app.config.get('SQLALCHEMY_BINDS')
            bind = binds.get(cls.__bind_key__)
        else:
            bind = current_app.config.get('SQLALCHEMY_DATABASE_URI')
        return urlparse(bind).path.strip('/')
    
    
    class T2(db.Model):
        __bind_key__ = 'db2'
        __tablename__ = "t2"
    
        @declared_attr
        def __table_args__(cls):
            return {'schema': _find_schema(cls)}
    
        t2_id = db.Column(db.Integer(), primary_key=True)
    
    
    class T1(db.Model):
        __tablename__ = "t1"
    
        @declared_attr
        def __table_args__(cls):
            return {'schema': _find_schema(cls)}

        t1_id = db.Column(db.Integer(), primary_key=True)
        t2_id = db.Column(db.Integer(), create_fk('t2.t2_id', schema=T2.__table_args__['schema']))
    
    
    if __name__ == '__main__':
        manager.run()

In short, I have a helper function `_find_schema()` to find the name of the schema based on the `__bind_key__` and the resulting value is set in `__table_args__`. I have a second helper function `create_fk()` to allow me to specify the name of the schema when creating the explicit name of the ForeignKey which is the expected format for alembic. Feels slightly messy but it does the job.

**Comment by miguelgrinberg** [comment]
Closing this, as it is not an issue with this project.
